interop-testing/src/main/java/io/grpc/testing/integration/AbstractInteropTest.java

```diff
@@ -1330,6 +1333,67 @@ public void perRpcCreds(String jsonKey, InputStream credentialsStream, String oa
     oauth2AuthToken(jsonKey, credentialsStream, oauthScope);
   }
 
+  /** Sends a cacheable unary rpc using GET. Requires that the server is behind a caching proxy. */
+  public void cacheableUnary() {
+    // Build method descriptor manually to set idempotent and safe to true.
+    io.grpc.MethodDescriptor<
+            io.grpc.testing.integration.Messages.SimpleRequest,
+            io.grpc.testing.integration.Messages.SimpleResponse>
+        safeCacheableUnaryCallMethod =
+            io.grpc.MethodDescriptor
+                .<io.grpc.testing.integration.Messages.SimpleRequest,
+                    io.grpc.testing.integration.Messages.SimpleResponse>
+                    newBuilder()
+                .setType(io.grpc.MethodDescriptor.MethodType.UNARY)
+                .setFullMethodName(
+                    MethodDescriptor.generateFullMethodName(
+                        "grpc.testing.TestService", "CacheableUnaryCall"))
+                .setRequestMarshaller(
+                    io.grpc.protobuf.ProtoUtils.marshaller(
+                        io.grpc.testing.integration.Messages.SimpleRequest.getDefaultInstance()))
+                .setResponseMarshaller(
+                    io.grpc.protobuf.ProtoUtils.marshaller(
+                        io.grpc.testing.integration.Messages.SimpleResponse.getDefaultInstance()))
+                .setIdempotent(true)
+                .setSafe(true)
+                .build();
+    // Set fake user IP since some proxies (GFE) won't cache requests from localhost.
+    Metadata.Key<String> userIpKey = Metadata.Key.of("x-user-ip", Metadata.ASCII_STRING_MARSHALLER);
+    Metadata metadata = new Metadata();
+    metadata.put(userIpKey, "1.2.3.4");
+    Channel channelWithUserIpKey =
+        ClientInterceptors.intercept(channel, MetadataUtils.newAttachHeadersInterceptor(metadata));
+    SimpleRequest requests1And2 =
+        SimpleRequest.newBuilder()
+            .setResponseSize(1)
+            .setResponseType(PayloadType.COMPRESSABLE)
+            .setPayload(
+                Payload.newBuilder()
+                    .setBody(ByteString.copyFromUtf8(String.valueOf(System.nanoTime()))))
+            .build();
+    SimpleRequest request3 =
+        SimpleRequest.newBuilder()
+            .setResponseSize(1)
+            .setResponseType(PayloadType.COMPRESSABLE)
+            .setPayload(
+                Payload.newBuilder()
+                    .setBody(ByteString.copyFromUtf8(String.valueOf(System.nanoTime()))))
+            .build();
+
+    SimpleResponse response1 =
+        ClientCalls.blockingUnaryCall(
+            channelWithUserIpKey, safeCacheableUnaryCallMethod, CallOptions.DEFAULT, requests1And2);
+    SimpleResponse response2 =
+        ClientCalls.blockingUnaryCall(
+            channelWithUserIpKey, safeCacheableUnaryCallMethod, CallOptions.DEFAULT, requests1And2);
+    SimpleResponse response3 =
```

Comment:
The interop test specification needs updating, which I was holding off on to do along with adding the verificiation of the payload as mentioned above. See https://github.com/grpc/grpc/blob/d93a8bead39ab112b808a248adb4576e4c9ce766/test/cpp/interop/interop_client.cc#L921 for the desired behavior.